src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.RelatedDocuments;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.RelatedDocuments;
+
+[ExportLanguageService(typeof(IRelatedDocumentsService), LanguageNames.CSharp), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CSharpRelatedDocumentsService() : AbstractRelatedDocumentsService
+{
+    protected override async ValueTask GetRelatedDocumentIdsInCurrentProcessAsync(
+        Document document,
+        int position,
+        Func<ImmutableArray<DocumentId>, CancellationToken, ValueTask> callbackAsync,
+        CancellationToken cancellationToken)
+    {
+        var solution = document.Project.Solution;
+
+        // Don't need nullable analysis, and we're going to walk a lot of the tree, so speed things up by not doing
+        // excess semantic work.
+        var semanticModel = await document.GetRequiredNullableDisabledSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        var seenDocumentIds = new ConcurrentSet<DocumentId>();
+        var seenTypeNames = new ConcurrentSet<string>();
+
+        // Bind as much as we can in parallel to get the results as quickly as possible.  We call into the
+        // ProducerConsumer overload that will batch up values into arrays, and call into the callback with them.  While
+        // that callback is executing, the ProducerConsumer continues to run, computing more results and getting them
+        // ready for when that call returns.  This approach ensures that we're not pausing while we're reporting the
+        // results to whatever client is calling into us.
+        await ProducerConsumer<DocumentId>.RunParallelAsync(
+            // Order the nodes by the distance from the requested position.
+            IteratePotentialTypeNodes(root).OrderBy(t => t.expression.SpanStart - position),
```

Comment:
> t.expression.SpanStart - position [](http://example.com/codeflow?start=56&length=34)

Should this be the absolute value of `t.expression.SpanStart - position` ?

Reply:
yup.  That would make more sense!